x/group/keeper/keeper_test.go

```diff
@@ -2416,22 +2422,206 @@ func (s *TestSuite) TestExecProposal() {
+			expProposalResult: group.PROPOSAL_RESULT_ACCEPTED,
+			expExecutorResult: group.PROPOSAL_EXECUTOR_RESULT_FAILURE,
+		},
+		"pruned when proposal is executable when failed before": {
+			setupProposal: func(ctx context.Context) uint64 {
+				msgs := []sdk.Msg{msgSend2}
+				myProposalID := submitProposalAndVote(ctx, s, msgs, proposers, group.VOTE_OPTION_YES)
+
+				_, err := s.keeper.Exec(ctx, &group.MsgExec{Signer: addr1.String(), ProposalId: myProposalID})
+				s.Require().NoError(err)
+				sdkCtx := sdk.UnwrapSDKContext(ctx)
+				s.Require().NoError(testutil.FundAccount(s.app.BankKeeper, sdkCtx, s.groupPolicyAddr, sdk.Coins{sdk.NewInt64Coin("test", 10002)}))
+
+				return myProposalID
+			},
+			expErrMsg:         "load proposal: not found",
+			expProposalStatus: group.PROPOSAL_STATUS_CLOSED,
+			expProposalResult: group.PROPOSAL_RESULT_ACCEPTED,
+			expExecutorResult: group.PROPOSAL_EXECUTOR_RESULT_SUCCESS,
+		},
+	}
+	for msg, spec := range specs {
+		spec := spec
+		s.Run(msg, func() {
+			sdkCtx, _ := s.sdkCtx.CacheContext()
+			ctx := sdk.WrapSDKContext(sdkCtx)
+			proposalID := spec.setupProposal(ctx)
+
+			if !spec.srcBlockTime.IsZero() {
+				sdkCtx = sdkCtx.WithBlockTime(spec.srcBlockTime)
+			}
+
+			ctx = sdk.WrapSDKContext(sdkCtx)
+			_, err := s.keeper.Exec(ctx, &group.MsgExec{Signer: addr1.String(), ProposalId: proposalID})
+			if spec.expErr {
+				s.Require().Error(err)
+				return
+			}
 			s.Require().NoError(err)
-			proposal := res.Proposal
 
-			exp := group.ProposalResult_name[int32(spec.expProposalResult)]
-			got := group.ProposalResult_name[int32(proposal.Result)]
-			s.Assert().Equal(exp, got)
+			if spec.expExecutorResult == group.PROPOSAL_EXECUTOR_RESULT_SUCCESS {
+				_, err := s.keeper.Proposal(ctx, &group.QueryProposalRequest{ProposalId: proposalID})
+				s.Require().Contains(err.Error(), spec.expErrMsg)
 
-			exp = group.ProposalStatus_name[int32(spec.expProposalStatus)]
-			got = group.ProposalStatus_name[int32(proposal.Status)]
-			s.Assert().Equal(exp, got)
+			} else {
+				// proposal is updated
+				res, err := s.keeper.Proposal(ctx, &group.QueryProposalRequest{ProposalId: proposalID})
+				s.Require().NoError(err)
+				proposal := res.Proposal
+				s.Require().Equal("", spec.expErrMsg)
+
+				exp := group.ProposalResult_name[int32(spec.expProposalResult)]
+				got := group.ProposalResult_name[int32(proposal.Result)]
+				s.Assert().Equal(exp, got)
 
-			exp = group.ProposalExecutorResult_name[int32(spec.expExecutorResult)]
-			got = group.ProposalExecutorResult_name[int32(proposal.ExecutorResult)]
-			s.Assert().Equal(exp, got)
+				exp = group.ProposalStatus_name[int32(spec.expProposalStatus)]
+				got = group.ProposalStatus_name[int32(proposal.Status)]
+				s.Assert().Equal(exp, got)
+
+				exp = group.ProposalExecutorResult_name[int32(spec.expExecutorResult)]
+				got = group.ProposalExecutorResult_name[int32(proposal.ExecutorResult)]
+				s.Assert().Equal(exp, got)
```

Comment:
In the else case, no need to check for values. Just check that proposal exists, and all votes exist